lms/views/admin/application_instance.py

```diff
@@ -9,17 +9,63 @@
 from lms.security import Permissions
 from lms.services import ApplicationInstanceNotFound, LTIRegistrationService
 from lms.validation._base import PyramidRequestSchema
-from lms.views.admin import flash_validation
+from lms.views.admin import error_render_to_response, flash_validation
+
+
+class EmptyStringNoneMixin:
+    """
+    Allows empty string as "missing value".
+
+    Marshmallow doesn't have a clean solution yet to POSTed values
+    (that are always present  in the request as empty strings)
+
+    Here we convert them explicitly to None
+
+    https://github.com/marshmallow-code/marshmallow/issues/713
+    """
+
+    def deserialize(self, value, attr, data, **kwargs):
+        if not value:
+            return None
+        return super().deserialize(value, attr, data, **kwargs)
+
+
+class EmptyStringURL(EmptyStringNoneMixin, fields.URL):
+    ...
+
+
+class EmptyStringEmail(EmptyStringNoneMixin, fields.Email):
+    ...
 
 
 class ApplicationInstanceSchema(PyramidRequestSchema):
     location = "form"
 
     deployment_id = fields.Str(required=True, validate=validate.Length(min=1))
+
+    developer_key = fields.Str(required=False, allow_none=True)
+    developer_secret = fields.Str(required=False, allow_none=True)
+
+
+class NewApplicationInstanceSchema(PyramidRequestSchema):
+    location = "form"
+
+    deployment_id = fields.Str(required=True, validate=validate.Length(min=1))
+
+    developer_key = fields.Str(required=False, allow_none=True)
+    developer_secret = fields.Str(required=False, allow_none=True)
+
     lms_url = fields.URL(required=True)
     email = fields.Email(required=True)
 
 
+class UpdateApplicationInstanceSchema(ApplicationInstanceSchema):
+    location = "form"
+
+    lms_url = EmptyStringURL(required=False, allow_none=True)
+    email = EmptyStringEmail(required=False, allow_none=True)
```

Comment:
Note to self: this looks like `NewApplicationInstanceSchema` could be a child of this, but it can't because the email and url fields have different missing behavior.